core/services/pipeline/orm.go

```diff
@@ -0,0 +1,360 @@
+		// Fetch the spec ID from the job
+		var job models.JobSpecV2
+		err = o.db.Where("id = ?", jobID).First(&job).Error
+		if err != nil {
+			return err
+		}
+
+		// Ensure the spec exists
+		var spec Spec
+		err = o.db.Where("id = ?", job.PipelineSpecID).First(&spec).Error
+		if err != nil {
+			return err
+		}
+
+		// Create the job run
+		run := Run{
+			PipelineSpecID: spec.ID,
+			Meta:           JSONSerializable{Val: meta},
+			CreatedAt:      time.Now(),
+		}
+		err = tx.Create(&run).Error
+		if err != nil {
+			return errors.Wrap(err, "could not create pipeline run")
+		}
+		runID = run.ID
+
+		// Create the task runs
+		err = tx.Exec(`
+            INSERT INTO pipeline_task_runs (
+                pipeline_run_id, pipeline_task_spec_id, index, dot_id, created_at
+            )
+            SELECT ? AS pipeline_run_id, id AS pipeline_task_spec_id, index, dot_id, NOW() AS created_at
+            FROM pipeline_task_specs
+            WHERE pipeline_spec_id = ?
+        `, run.ID, run.PipelineSpecID).Error
+		return errors.Wrap(err, "could not create pipeline task runs")
+	})
+	return runID, err
+}
+
+// ProcessNextUnclaimedTaskRun chooses any arbitrary incomplete TaskRun from the DB
+// whose parent TaskRuns have already been processed.
+func (o *orm) ProcessNextUnclaimedTaskRun(fn func(jobID int32, ptRun TaskRun, predecessors []TaskRun) Result) (_ bool, err error) {
+	var done bool
+	err = utils.GormTransaction(o.db, func(tx *gorm.DB) (err error) {
+		var ptRun TaskRun
+		var predecessors []TaskRun
+
+		// Find the next unlocked, unfinished pipeline_task_run with no uncompleted predecessors
+		err = tx.Raw(`
+            SELECT * from pipeline_task_runs WHERE id IN (
+                SELECT pipeline_task_runs.id FROM pipeline_task_runs
+                    INNER JOIN pipeline_task_specs ON pipeline_task_runs.pipeline_task_spec_id = pipeline_task_specs.id
+                    LEFT JOIN pipeline_task_specs AS predecessor_specs ON predecessor_specs.successor_id = pipeline_task_specs.id
+                    LEFT JOIN pipeline_task_runs AS predecessor_unfinished_runs ON predecessor_specs.id = predecessor_unfinished_runs.pipeline_task_spec_id
+                          AND pipeline_task_runs.pipeline_run_id = predecessor_unfinished_runs.pipeline_run_id
+                WHERE pipeline_task_runs.finished_at IS NULL
+                GROUP BY (pipeline_task_runs.id)
+                HAVING (
+                    bool_and(predecessor_unfinished_runs.finished_at IS NOT NULL)
+                    OR
+                    count(predecessor_unfinished_runs.id) = 0
+                )
+            )
+            LIMIT 1
+            FOR UPDATE OF pipeline_task_runs SKIP LOCKED;
+        `).Scan(&ptRun).Error
+		if gorm.IsRecordNotFoundError(err) {
+			done = true
+			return nil
+		} else if err != nil {
+			return errors.Wrap(err, "error finding next task run")
+		}
+
+		// Fill in the TaskSpec
```

Comment:
Do you mean PipelineTaskSpec. The inconsistent naming is confusing

Reply:
The way I've been thinking about it is that `pipeline.TaskSpec` is the struct and `PipelineTaskSpec` is the field on other structs that refers to it.  `pipeline.PipelineTaskSpec` seems suboptimal.  But I'll add the package prefix to this comment for clarity